src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractCreateServicesOnTextViewConnection.cs

```diff
@@ -15,20 +14,18 @@
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
 {
     /// <summary>
-    /// Ensures services that must be constructed on the UI thread are appropriately created during
-    /// the first connection of an applicable subject buffer to an IWpfTextView. This ensures the
-    /// services are available by the time an open document or the interactive window needs them.
-    /// The <see cref="CreateServicesOnUIThread(IComponentModel, string)"/> method should also be
+    /// Creates services on the first connection of an applicable subject buffer to an IWpfTextView. 
+    /// This ensures the services are available by the time an open document or the interactive window needs them.
+    /// The <see cref="CreateServices(IComponentModel, string)"/> method should also be
     /// called during package load to front load some of the work.
     /// </summary>
-    internal abstract class HACK_AbstractCreateServicesOnUiThread : ForegroundThreadAffinitizedObject, IWpfTextViewConnectionListener
+    internal abstract class AbstractCreateServicesOnTextViewConnection : IWpfTextViewConnectionListener
     {
         private readonly IComponentModel _componentModel;
         private readonly string _languageName;
         private bool _initialized = false;
 
-        public HACK_AbstractCreateServicesOnUiThread(IThreadingContext threadingContext, IServiceProvider serviceProvider, string languageName)
-            : base(threadingContext)
+        public AbstractCreateServicesOnTextViewConnection(IServiceProvider serviceProvider, string languageName)
         {
             _componentModel = (IComponentModel)serviceProvider.GetService(typeof(SComponentModel));
```

Comment:
Wait, so this thing gets the IServiceProvider only to go back and get the component model from it? Can we write this with just MEF directly?

Reply:
I attempted this - I could not import IComponentModel (composition failed) and did not find anything directly importing it from mef anywhere else.  If you have another way let me know!